### ISSUE TITLE
Old version pods are created when deploying a new version with more replicas

### ISSUE THREAD

**Issue by prasus** [kubernetes/kubernetes#105395]
#### What happened:

We recently noticed a weird behavior as explained below in our Kubernetes cluster.

`#1`. The initial version of the deployment is running with **2 replicas** as seen below.

```
$ kubectl -n like-prediction get pods
NAME                                 READY   STATUS            RESTARTS   AGE
user-likes-7c4b4cb947-874v8          2/2     Running           0          2m10s
user-likes-7c4b4cb947-ctddb          2/2     Running           0          2m41s
```
`#2`. Then we deploy an updated version with **3 replicas**, during the deployment process it does something surprising, Instead of only spinning up pods of the new version with the desired number of replicas and terminating pods of the old version, **a third pod of the old version is created (and then terminated)** as seen below.

```
$ kubectl -n like-prediction get pods
NAME                                 READY   STATUS            RESTARTS   AGE
user-likes-5cdc8f4f55-zz9vf          0/2     Running           0          4s
user-likes-7c4b4cb947-874v8          2/2     Running           0          4m26s
user-likes-7c4b4cb947-ctddb          2/2     Running           0          4m57s
user-likes-7c4b4cb947-r79k4          0/2     PodInitializing   0          4s

$ kubectl -n like-prediction get pods
NAME                                 READY   STATUS    RESTARTS   AGE
user-likes-5cdc8f4f55-zz9vf          1/2     Running   0          19s
user-likes-7c4b4cb947-874v8          2/2     Running   0          4m41s
user-likes-7c4b4cb947-ctddb          2/2     Running   0          5m12s
user-likes-7c4b4cb947-r79k4          1/2     Running   0          19s

$ kubectl -n like-prediction get pods
NAME                                 READY   STATUS        RESTARTS   AGE
user-likes-5cdc8f4f55-99mmj          0/2     Init:0/1      0          3s
user-likes-5cdc8f4f55-zz9vf          2/2     Running       0          34s
user-likes-7c4b4cb947-874v8          2/2     Running       0          4m56s
user-likes-7c4b4cb947-ctddb          2/2     Running       0          5m27s
user-likes-7c4b4cb947-r79k4          1/2     Terminating   0          34s
```
```
snip from kubernetes namespace events:

69s         Normal    ScalingReplicaSet   deployment/user-likes                              Scaled up replica set user-likes-7c4b4cb947 to 3
69s         Normal    ScalingReplicaSet   deployment/user-likes                              Scaled up replica set user-likes-5cdc8f4f55 to 1
15m         Normal    ScalingReplicaSet   deployment/user-likes                              Scaled down replica set user-likes-7c4b4cb947 to 2
38s         Normal    ScalingReplicaSet   deployment/user-likes                              Scaled up replica set user-likes-5cdc8f4f55 to 2
17s         Normal    ScalingReplicaSet   deployment/user-likes                              Scaled down replica set user-likes-7c4b4cb947 to 1
38s         Normal    ScalingReplicaSet   deployment/user-likes                              (combined from similar events): Scaled down replica set user-likes-7c4b4cb947 to 2
17s         Normal    ScalingReplicaSet   deployment/user-likes                              Scaled up replica set user-likes-5cdc8f4f55 to 3

controller logs:

I1001 06:58:04.621233       1 event.go:291] "Event occurred" object="like-prediction/user-likes" kind="Deployment" apiVersion="apps/v1" type="Normal" reason="ScalingReplicaSet" message="Scaled up replica set user-likes-7c4b4cb947 to 3"
I1001 06:58:04.646022       1 event.go:291] "Event occurred" object="like-prediction/user-likes" kind="Deployment" apiVersion="apps/v1" type="Normal" reason="ScalingReplicaSet" message="Scaled up replica set user-likes-5cdc8f4f55 to 1"
I1001 06:58:04.703671       1 event.go:291] "Event occurred" object="like-prediction/user-likes-7c4b4cb947" kind="ReplicaSet" apiVersion="apps/v1" type="Normal" reason="SuccessfulCreate" message="Created pod: user-likes-7c4b4cb947-r79k4"
I1001 06:58:04.721249       1 event.go:291] "Event occurred" object="like-prediction/user-likes-5cdc8f4f55" kind="ReplicaSet" apiVersion="apps/v1" type="Normal" reason="SuccessfulCreate" message="Created pod: user-likes-5cdc8f4f55-zz9vf"
I1001 06:58:35.644563       1 event.go:291] "Event occurred" object="like-prediction/user-likes-7c4b4cb947" kind="ReplicaSet" apiVersion="apps/v1" type="Normal" reason="SuccessfulDelete" message="Deleted pod: user-likes-7c4b4cb947-r79k4"
I1001 06:58:35.647837       1 event.go:291] "Event occurred" object="like-prediction/user-likes" kind="Deployment" apiVersion="apps/v1" type="Normal" reason="ScalingReplicaSet" message="Scaled down replica set user-likes-7c4b4cb947 to 2"
I1001 06:58:35.751554       1 event.go:291] "Event occurred" object="like-prediction/user-likes" kind="Deployment" apiVersion="apps/v1" type="Normal" reason="ScalingReplicaSet" message="Scaled up replica set user-likes-5cdc8f4f55 to 2"
I1001 06:58:35.899229       1 event.go:291] "Event occurred" object="like-prediction/user-likes-5cdc8f4f55" kind="ReplicaSet" apiVersion="apps/v1" type="Normal" reason="SuccessfulCreate" message="Created pod: user-likes-5cdc8f4f55-99mmj"
I1001 06:58:56.312886       1 event.go:291] "Event occurred" object="like-prediction/user-likes" kind="Deployment" apiVersion="apps/v1" type="Normal" reason="ScalingReplicaSet" message="Scaled down replica set user-likes-7c4b4cb947 to 1"
I1001 06:58:56.351618       1 event.go:291] "Event occurred" object="like-prediction/user-likes-7c4b4cb947" kind="ReplicaSet" apiVersion="apps/v1" type="Normal" reason="SuccessfulDelete" message="Deleted pod: user-likes-7c4b4cb947-874v8"
I1001 06:58:56.367960       1 event.go:291] "Event occurred" object="like-prediction/user-likes" kind="Deployment" apiVersion="apps/v1" type="Normal" reason="ScalingReplicaSet" message="Scaled up replica set user-likes-5cdc8f4f55 to 3"
I1001 06:58:56.464972       1 event.go:291] "Event occurred" object="like-prediction/user-likes-5cdc8f4f55" kind="ReplicaSet" apiVersion="apps/v1" type="Normal" reason="SuccessfulCreate" message="Created pod: user-likes-5cdc8f4f55-h8n9w"
I1001 06:59:46.717529       1 event.go:291] "Event occurred" object="like-prediction/user-likes" kind="Deployment" apiVersion="apps/v1" type="Normal" reason="ScalingReplicaSet" message="Scaled down replica set user-likes-7c4b4cb947 to 0"
I1001 06:59:46.751150       1 event.go:291] "Event occurred" object="like-prediction/user-likes-7c4b4cb947" kind="ReplicaSet" apiVersion="apps/v1" type="Normal" reason="SuccessfulDelete" message="Deleted pod: user-likes-7c4b4cb947-ctddb"
```
`#3` After a couple of minutes everything looks good as expected. 

```
$ kubectl -n like-prediction get pods
NAME                                 READY   STATUS    RESTARTS   AGE
user-likes-5cdc8f4f55-99mmj          2/2     Running   0          2m4s
user-likes-5cdc8f4f55-h8n9w          2/2     Running   0          103s
user-likes-5cdc8f4f55-zz9vf          2/2     Running   0          2m35s
```

As we see from the above steps, kubectl command's and the logs, the new version deployment with **3 replicas** initially scales up the `replicaSet` of the old version to **3** and then terminates it, this is abnormal (or something we haven't experienced so far with our K8s infrastructure), This is not much of a probelm as things automatically gets corrected eventually but I am really curious to know more, Is this an intended behavior, or a bug? 

#### What you expected to happen:

Ideally, the updated version deployment process will spin up pods of the new version with the desired number of replicas and terminates the pods of the old version. 

#### How to reproduce it (as minimally and precisely as possible):

I am always able to reproduce this behavior on our Kubernetes cluster running **v1.21.4** by following the steps denoted below.

`#1`. Provision a simple nginx deployment with 2 replicas using the below manifest.

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
  labels:
    release: testdeploy
    app: nginx
spec:
  replicas: 2
  strategy:
    type: RollingUpdate
    rollingUpdate:
      maxSurge: 1
      maxUnavailable: 30%
  selector:
    matchLabels:
      release: testdeploy
      app: nginx
  template:
    metadata:
      labels:
        release: testdeploy
        app: nginx
      annotations:
    spec:
      imagePullSecrets:
      tolerations:
      affinity:
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
          - weight: 100
            podAffinityTerm:
              topologyKey: "kubernetes.io/hostname"
              labelSelector:
                matchLabels:
                  release: testdeploy
                  app: nginx
      containers:
        - name: web
          image: "nginx:1.20"
```
The deployment works as expected. 

```
 kubectl -n test-elasticsearch get pods
NAME                     READY   STATUS    RESTARTS   AGE
nginx-844bf5fdb8-vn72z   1/1     Running   0          15s
nginx-844bf5fdb8-w45ww   1/1     Running   0          15s
```

`#2`.  Update the deployment by just changing the image version and also increase the replicas to 3 as shown below

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
  labels:
    release: testdeploy
    app: nginx
spec:
  replicas: 3
  strategy:
    type: RollingUpdate
    rollingUpdate:
      maxSurge: 1
      maxUnavailable: 30%
  selector:
    matchLabels:
      release: testdeploy
      app: nginx
  template:
    metadata:
      labels:
        release: testdeploy
        app: nginx
      annotations:
    spec:
      imagePullSecrets:
      tolerations:
      affinity:
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
          - weight: 100
            podAffinityTerm:
              topologyKey: "kubernetes.io/hostname"
              labelSelector:
                matchLabels:
                  release: testdeploy
                  app: nginx
      containers:
        - name: web
          image: "nginx:1.21"
```

We can see that the deployment update with **3 replicas** will initially scales up the `replicaSet` of the old version to **3** and then terminates it after a few seconds.

```
kubectl -n test-elasticsearch get pods

NAME                     READY   STATUS              RESTARTS   AGE
nginx-6bff497f57-42bj5   0/1     ContainerCreating   0          1s
nginx-6bff497f57-st6tp   1/1     Running             0          4s
nginx-844bf5fdb8-894dp   0/1     Terminating         0          4s
nginx-844bf5fdb8-vn72z   1/1     Running             0          94s
nginx-844bf5fdb8-w45ww   1/1     Running             0          94s

namespace event logs

88s         Normal    ScalingReplicaSet   deployment/nginx                   Scaled up replica set nginx-844bf5fdb8 to 3
88s         Normal    ScalingReplicaSet   deployment/nginx                   Scaled up replica set nginx-6bff497f57 to 1
85s         Normal    ScalingReplicaSet   deployment/nginx                   Scaled down replica set nginx-844bf5fdb8 to 2
85s         Normal    ScalingReplicaSet   deployment/nginx                   Scaled up replica set nginx-6bff497f57 to 2
76s         Normal    ScalingReplicaSet   deployment/nginx                   Scaled down replica set nginx-844bf5fdb8 to 1
76s         Normal    ScalingReplicaSet   deployment/nginx                   Scaled up replica set nginx-6bff497f57 to 3
73s         Normal    ScalingReplicaSet   deployment/nginx                   Scaled down replica set nginx-844bf5fdb8 to 0
```

#### Anything else we need to know?:

We were on Kubernetes **v1.18.8** before and have upgraded the environment to **v1.21.4** a few weeks back, We have never seen this happening before the upgrade. 

#### Environment:
- Kubernetes version (use `kubectl version`): `v1.21.4`
- Cloud provider or hardware configuration: `Self hosted Kubernetes, Deployed using kubeadm`.
- OS (e.g: `cat /etc/os-release`): `Ubuntu Bionic (Ubuntu 18.04.5 LTS)`
- Kernel (e.g. `uname -a`): `4.15.0-115-generic`
- Install tools: `kubeadm`
- Network plugin and version (if this is a network-related bug): 
- Others:

**Comment by JaneLiuL** [comment]
I just check the code like below flow(please correct me if I am wrong):
1. Check is it need to scale, since it is, so scale to 3
2. Update the replicaset and the pods
```
       // first will check is it need to scale:  compare the rs.Spec.Replicas) == deployment.Spec.Replicas or not
	scalingEvent, err := dc.isScalingEvent(d, rsList)

       // if yes, then will execute the `scaleReplicaSetAndRecordEvent` to scale
	if scalingEvent {
		return dc.sync(d, rsList)
	}

      // Since the first time already scale rs to the same, then will continue below part, start to update the replicaset and the pods
	switch d.Spec.Strategy.Type {
	case apps.RecreateDeploymentStrategyType:
		return dc.rolloutRecreate(d, rsList, podMap)
	case apps.RollingUpdateDeploymentStrategyType:
		return dc.rolloutRolling(d, rsList)
	}

}
```

**Comment by prasus** [comment]
Is there an update on this? 🙁

**Comment by RyanAoh** [comment]
@prasus  Uh, I've found the codes which lead to this issue. 
But I'm a little busy recently, I 'll submit a PR asap.

**Comment by prasus** [comment]
Thanks for the update, @RyanAoh 👍🏿

**Comment by RyanAoh** [comment]
According to the source code of the function [`(dc *DeploymentController) scale(ctx context.Context...)`](https://github.com/kubernetes/kubernetes/blob/83ca74541216405323ddfb67f5f80ad5717da826/pkg/controller/deployment/sync.go#L298), the scaling of the replicaset with old spec is expected.
Refer to these coments：https://github.com/kubernetes/kubernetes/blob/83ca74541216405323ddfb67f5f80ad5717da826/pkg/controller/deployment/sync.go#L320-L322

@prasus Did these old version pods have any negative effects in your case？

**Comment by nagygergo** [comment]
/reopen
/remove-lifecycle rotten

We've observed this issue and it was causing issues when we wanted to change the version of the software deployed while also scaling out the deployment. We have some data store configuration and database schema update scripts implemented as init containers. In our case needed to do a non backward compatible configuration upgrade, so the old version of the service was not able to work with the new config, and the new version of the service was not able to work with the new config. 

The following happened:
1. Updated deployment to have 2 replicas instead of 1. (default maxUnavailable and maxSurge configs)
2. New ReplicaSet created with 1 podv2. Old ReplicaSet scaled up to 2 podv1-s.
3. podv2 updated data store config.
4. new podv1 updated data store config.
5. Eventually podv2 failed liveness probe, because couldn't contact data store correctly.
6. RollingUpdate never finished.
